File: Fieldframe/shared/src/commonMain/kotlin/com/fieldframe/shared/export/PhotoDoc.kt
package com.fieldframe.shared.export

import com.fieldframe.shared.domain.DEFAULT_SECTION_NAMES
import com.fieldframe.shared.domain.currentTimeMillis
import com.fieldframe.shared.domain.extensionFor
import com.fieldframe.shared.domain.newId
import com.fieldframe.shared.domain.sanitizeFilename
import com.fieldframe.shared.model.Photo
import com.fieldframe.shared.model.Project
import com.fieldframe.shared.model.ProjectBundle
import com.fieldframe.shared.model.Section

data class PhotoDocPackage(
    val bundle: ProjectBundle,
    val photoBytes: Map<String, ByteArray>,
)

data class ExportedPackage(
    val bytes: ByteArray,
    val filename: String,
) {
    override fun equals(other: Any?): Boolean =
        other is ExportedPackage && bytes.contentEquals(other.bytes) && filename == other.filename

    override fun hashCode(): Int = 31 * bytes.contentHashCode() + filename.hashCode()
}

fun packageFilename(projectName: String): String = "${sanitizeFilename(projectName)}.photodoc"

fun encodePhotoDocJson(bundle: ProjectBundle): String {
    val p = bundle.project
    val sections = bundle.sections.joinToString(",") { s ->
        """{"id":${j(s.id)},"projectId":${j(s.projectId)},"name":${j(s.name)},"sortOrder":${s.sortOrder},"isCustom":${s.isCustom}}"""
    }
    val photos = bundle.photos.joinToString(",") { ph ->
        """{"id":${j(ph.id)},"projectId":${j(ph.projectId)},"sectionId":${j(ph.sectionId)},"description":${j(ph.description)},"sortOrder":${ph.sortOrder},"createdAt":${ph.createdAt},"updatedAt":${ph.updatedAt},"rotation":${ph.rotation},"mimeType":${j(ph.mimeType)},"width":${ph.width},"height":${ph.height},"phase":${j(ph.phase)},"originalName":${j(ph.originalName)}}"""
    }
    return """{"format":"photodoc","version":1,"exportedAt":${j(isoNow())},"project":{"id":${j(p.id)},"name":${j(p.name)},"description":${j(p.description)},"date":${j(p.date)},"referenceNumber":${j(p.referenceNumber)},"createdAt":${p.createdAt},"updatedAt":${p.updatedAt},"beforeAfterEnabled":${p.beforeAfterEnabled}},"sections":[$sections],"photos":[$photos]}"""
}

fun parsePhotoDoc(
    json: String,
    files: Map<String, ByteArray>,
): PhotoDocPackage {
    val root = JsonMap.parse(json)
    if (root.str("format") != "photodoc") {
        error("This file is not a valid Fieldframe package.")
    }
    val src = root.map("project") ?: error("The package is missing required project data.")
    val now = currentTimeMillis()
    val projectId = newId()
    val project = Project(
        id = projectId,
        name = src.str("name").trim().ifEmpty { "Imported project" },
        description = src.str("description"),
        date = src.str("date").ifEmpty { com.fieldframe.shared.domain.todayIsoDate() },
        referenceNumber = src.str("referenceNumber"),
        createdAt = now,
        updatedAt = now,
        beforeAfterEnabled = src.bool("beforeAfterEnabled"),
    )
    val sectionMap = HashMap<String, String>()
    val sections = root.list("sections").map { raw ->
        val id = newId()
        sectionMap[raw.str("id")] = id
        Section(
            id = id,
            projectId = projectId,
            name = raw.str("name").ifEmpty { "Section" },
            sortOrder = raw.int("sortOrder"),
            isCustom = raw.bool("isCustom"),
        )
    }.ifEmpty {
        DEFAULT_SECTION_NAMES.mapIndexed { index, name ->
            Section(newId(), projectId, name, index, false)
        }
    }
    val photos = ArrayList<Photo>()
    val bytes = HashMap<String, ByteArray>()
    for (raw in root.list("photos")) {
        val oldId = raw.str("id")
        val sectionId = sectionMap[raw.str("sectionId")] ?: continue
        val id = newId()
        val ext = extensionFor(raw.str("mimeType"), raw.str("originalName"))
        val file = files["photos/$oldId.$ext"]
            ?: files.entries.firstOrNull { it.key.startsWith("photos/$oldId.") }?.value
            ?: continue
        photos += Photo(
            id = id,
            projectId = projectId,
            sectionId = sectionId,
            description = raw.str("description"),
            sortOrder = raw.int("sortOrder"),
            createdAt = now,
            updatedAt = now,
            rotation = raw.int("rotation"),
            mimeType = raw.str("mimeType").ifEmpty { "image/jpeg" },
            width = raw.int("width"),
            height = raw.int("height"),
            phase = raw.str("phase").ifEmpty { "standard" },
            originalName = raw.str("originalName").ifEmpty { "photo-$id.jpg" },
        )
        bytes[id] = file
    }
    return PhotoDocPackage(ProjectBundle(project, sections, photos), bytes)
}

fun zipPhotoDoc(bundle: ProjectBundle, photoBytes: Map<String, ByteArray>): ByteArray {
    val entries = LinkedHashMap<String, ByteArray>()
    entries["project.json"] = encodePhotoDocJson(bundle).encodeToByteArray()
    for (photo in bundle.photos) {
        val data = photoBytes[photo.id] ?: continue
        val ext = extensionFor(photo.mimeType, photo.originalName)
        entries["photos/${photo.id}.$ext"] = data
    }
    return ZipStore.encode(entries)
}

fun unzipPhotoDoc(bytes: ByteArray): PhotoDocPackage {
    val entries = ZipStore.decode(bytes)
    val json = entries["project.json"]?.decodeToString()
        ?: error("This file is not a valid Fieldframe package.")
    return parsePhotoDoc(json, entries)
}

private fun j(value: String): String = buildString {
    append('"')
    for (ch in value) {
        when (ch) {
            '\\' -> append("\\\\")
            '"' -> append("\\\"")
            '\n' -> append("\\n")
            '\r' -> append("\\r")
            '\t' -> append("\\t")
            else -> if (ch.code < 32) append("\\u").append(ch.code.toString(16).padStart(4, '0')) else append(ch)
        }
    }
    append('"')
}

private fun isoNow(): String {
    val ms = currentTimeMillis()
    val days = ms / 86_400_000L
    val rem = (ms % 86_400_000L).toInt()
    val h = rem / 3_600_000
    val m = (rem % 3_600_000) / 60_000
    val s = (rem % 60_000) / 1000
    val ymd = com.fieldframe.shared.domain.todayIsoDate(ms)
    return "${ymd}T${pad(h)}:${pad(m)}:${pad(s)}Z"
}

private fun pad(n: Int) = if (n < 10) "0$n" else "$n"

File: src/lib/export/photodoc.ts
import JSZip from "jszip";
import { z } from "zod";
import { getPhotoBlob, getPhotoThumb } from "../repository";
import {
  PHOTODOC_FORMAT,
  PHOTODOC_VERSION,
  type Photo,
  type PhotoDocManifest,
  type PhotoPhase,
  type Project,
  type Rotation,
  type Section,
} from "../models";
import { putProjectGraph } from "../idb";
import { makeThumbnail } from "../images";
import { newId, sanitizeFilename } from "../utils";

const rotationSchema = z.union([z.literal(0), z.literal(90), z.literal(180), z.literal(270)]);
const phaseSchema = z.union([z.literal("standard"), z.literal("before"), z.literal("after")]);

const manifestSchema = z.object({
  format: z.literal(PHOTODOC_FORMAT),
  version: z.literal(PHOTODOC_VERSION),
  exportedAt: z.string(),
  project: z.object({
    id: z.string(),
    name: z.string(),
    description: z.string(),
    date: z.string(),
    referenceNumber: z.string(),
    createdAt: z.number(),
    updatedAt: z.number(),
    beforeAfterEnabled: z.boolean(),
  }),
  sections: z.array(
    z.object({
      id: z.string(),
      projectId: z.string(),
      name: z.string(),
      sortOrder: z.number(),
      isCustom: z.boolean(),
    }),
  ),
  photos: z.array(
    z.object({
      id: z.string(),
      projectId: z.string(),
      sectionId: z.string(),
      description: z.string(),
      sortOrder: z.number(),
      createdAt: z.number(),
      updatedAt: z.number(),
      rotation: rotationSchema,
      mimeType: z.string(),
      width: z.number(),
      height: z.number(),
      phase: phaseSchema,
      originalName: z.string(),
    }),
  ),
});

function extensionFor(mime: string, name: string): string {
  const fromName = name.split(".").pop()?.toLowerCase();
  if (fromName && fromName.length <= 5) return fromName;
  if (mime.includes("png")) return "png";
  if (mime.includes("webp")) return "webp";
  if (mime.includes("heic") || mime.includes("heif")) return "heic";
  return "jpg";
}

export async function exportProjectPackage(bundle: {
  project: Project;
  sections: Section[];
  photos: Photo[];
}): Promise<{ blob: Blob; filename: string }> {
  const zip = new JSZip();
  const manifest: PhotoDocManifest = {
    format: PHOTODOC_FORMAT,
    version: PHOTODOC_VERSION,
    exportedAt: new Date().toISOString(),
    project: bundle.project,
    sections: bundle.sections,
    photos: bundle.photos,
  };
  zip.file("project.json", JSON.stringify(manifest, null, 2));
  const folder = zip.folder("photos");
  if (!folder) throw new Error("Could not create the export package.");
  for (const photo of bundle.photos) {
    const blob = await getPhotoBlob(photo.id);
    if (!blob) continue;
    const ext = extensionFor(photo.mimeType, photo.originalName);
    folder.file(`${photo.id}.${ext}`, blob);
  }
  const blob = await zip.generateAsync({ type: "blob" });
  const filename = `${sanitizeFilename(bundle.project.name)}.photodoc`;
  return { blob, filename };
}

export async function importProjectPackage(file: Blob): Promise<Project> {
  const zip = await JSZip.loadAsync(file);
  const jsonFile = zip.file("project.json");
  if (!jsonFile) throw new Error("This file is not a valid Fieldframe package.");
  let parsed: unknown;
  try {
    parsed = JSON.parse(await jsonFile.async("string"));
  } catch {
    throw new Error("The package manifest is corrupted.");
  }
  const result = manifestSchema.safeParse(parsed);
  if (!result.success) {
    throw new Error("The package is missing required project data.");
  }
  const manifest = result.data;
  const now = Date.now();
  const projectId = newId();
  const sectionMap = new Map<string, string>();
  const photoMap = new Map<string, string>();

  const project: Project = {
    ...manifest.project,
    id: projectId,
    name: manifest.project.name?.trim() || "Imported project",
    createdAt: now,
    updatedAt: now,
  };
  const sections: Section[] = manifest.sections.map((section) => {
    const id = newId();
    sectionMap.set(section.id, id);
    return { ...section, id, projectId };
  });
  const photos: Photo[] = [];
  const blobs: { id: string; blob: Blob }[] = [];
  const thumbs: { id: string; blob: Blob }[] = [];

  for (const photo of manifest.photos) {
    const sectionId = sectionMap.get(photo.sectionId);
    if (!sectionId) continue;
    const id = newId();
    photoMap.set(photo.id, id);
    const ext = extensionFor(photo.mimeType, photo.originalName);
    const entry =
      zip.file(`photos/${photo.id}.${ext}`) ||
      zip.file(`photos/${photo.id}.jpg`) ||
      zip.file(`photos/${photo.id}.jpeg`) ||
      zip.file(`photos/${photo.id}.png`);
    if (!entry) continue;
    const bytes = await entry.async("blob");
    const blob = bytes.type ? bytes : new Blob([bytes], { type: photo.mimeType || "image/jpeg" });
    let thumb: Blob;
    try {
      thumb = await makeThumbnail(blob, photo.rotation as Rotation);
    } catch {
      continue;
    }
    photos.push({
      ...photo,
      id,
      projectId,
      sectionId,
      phase: photo.phase as PhotoPhase,
      rotation: photo.rotation as Rotation,
    });
    blobs.push({ id, blob });
    thumbs.push({ id, blob: thumb });
  }

  if (sections.length === 0) {
    throw new Error("The package does not contain any sections.");
  }

  await putProjectGraph({ project, sections, photos, blobs, thumbs });
  return project;
}

export { getPhotoThumb };
